src/test/txdownload_tests.cpp

```diff
@@ -0,0 +1,248 @@
+
+        // Parent reconsiderable failure
+        const auto& [p1_keep, p1_unique_txids, p1_package] = txdownload_impl.MempoolRejectedTx(ptx_parent_1, state_reconsiderable, nodeid, /*first_time_failure=*/true);
+        BOOST_CHECK(p1_keep);
+        BOOST_CHECK(p1_unique_txids.empty());
+        BOOST_CHECK(!p1_package.has_value());
+
+        const auto& [p2_keep, p2_unique_txids, p2_package] = txdownload_impl.MempoolRejectedTx(ptx_parent_2, state_reconsiderable, nodeid, /*first_time_failure=*/true);
+        BOOST_CHECK(p2_keep);
+        BOOST_CHECK(p2_unique_txids.empty());
+        BOOST_CHECK(!p2_package.has_value());
+
+        // Child missing inputs. Not added to orphanage because only 1 reconsiderable rejected parent is tolerated.
+        const auto& [c_keep, c_unique_txids, c_package] = txdownload_impl.MempoolRejectedTx(ptx_child, state_orphan, nodeid, /*first_time_failure=*/true);
+        BOOST_CHECK(c_keep);
+        BOOST_CHECK(c_unique_txids == vec_parent_txids);
+        BOOST_CHECK(!c_package.has_value());
+        BOOST_CHECK(!txdownload_impl.m_orphanage.HaveTx(ptx_child->GetWitnessHash()));
+    }
+
+    // 1 reconsiderable parent + 1 non-reconsiderable parent
+    {
+        node::TxDownloadImpl txdownload_impl{node::TxDownloadOptions{pool, det_rand, DEFAULT_MAX_ORPHAN_TRANSACTIONS}};
+
+        // Parent reconsiderable failure
+        const auto& [p1_keep, p1_unique_txids, p1_package] = txdownload_impl.MempoolRejectedTx(ptx_parent_1, state_reconsiderable, nodeid, /*first_time_failure=*/true);
+        BOOST_CHECK(p1_keep);
+        BOOST_CHECK(p1_unique_txids.empty());
+        BOOST_CHECK(!p1_package.has_value());
+
+        TxValidationState state_not_reconsiderable;
+        state_not_reconsiderable.Invalid(TxValidationResult::TX_CONSENSUS, "");
+
+        const auto& [p2_keep, p2_unique_txids, p2_package] = txdownload_impl.MempoolRejectedTx(ptx_parent_2, state_not_reconsiderable, nodeid, /*first_time_failure=*/true);
+        BOOST_CHECK(p2_keep);
+        BOOST_CHECK(p2_unique_txids.empty());
+        BOOST_CHECK(!p2_package.has_value());
+
+        // Child missing inputs. Not added to orphanage because only 1 reconsiderable rejected parent is tolerated.
+        const auto& [c_keep, c_unique_txids, c_package] = txdownload_impl.MempoolRejectedTx(ptx_child, state_orphan, nodeid, /*first_time_failure=*/true);
+        BOOST_CHECK(c_keep);
+        BOOST_CHECK(c_unique_txids == vec_parent_txids);
+        BOOST_CHECK(!c_package.has_value());
+        BOOST_CHECK(!txdownload_impl.m_orphanage.HaveTx(ptx_child->GetWitnessHash()));
+
+    }
+
+    // Reconsiderable parent of a tx already in orphanage: the only time PackageToValidate is returned.
+    {
+        node::TxDownloadImpl txdownload_impl{node::TxDownloadOptions{pool, det_rand, DEFAULT_MAX_ORPHAN_TRANSACTIONS}};
+
+        // Child missing inputs, should be added to orphanage.
```

Comment:
nit: just assert it's in orphanage instead

Reply:
added